identity-iota/src/tangle/message/message_ext.rs

```diff
@@ -0,0 +1,191 @@
+
+// TODO: Use MessageId when it has a const ctor
+static NULL: &[u8; MESSAGE_ID_LENGTH] = &[0; MESSAGE_ID_LENGTH];
+
+fn parse_message<T: FromJson + TangleRef>(message: &Message, did: &IotaDID) -> Option<T> {
+  let message_id: MessageId = message.id().0;
+  let payload: Option<&Payload> = message.payload().as_ref();
+  let resource: T = parse_payload(message_id, payload)?;
+
+  if did.authority() != resource.did().authority() {
+    return None;
+  }
+
+  Some(resource)
+}
+
+fn parse_payload<T: FromJson + TangleRef>(message_id: MessageId, payload: Option<&Payload>) -> Option<T> {
+  match payload {
+    Some(Payload::Indexation(indexation)) => parse_data(message_id, indexation.data()),
+    Some(Payload::Transaction(transaction)) => match transaction.essence() {
+      Essence::Regular(essence) => match essence.payload() {
+        Some(Payload::Indexation(payload)) => parse_data(message_id, payload.data()),
+        _ => None,
+      },
+    },
+    _ => None,
+  }
+}
+
+// TODO: allow this to return errors?
+fn parse_data<T: FromJson + TangleRef>(message_id: MessageId, data: &[u8]) -> Option<T> {
+  // Check version.
+  let version: DIDMessageVersion = DIDMessageVersion::try_from(*data.get(0)?).ok()?;
+  if version != DIDMessageVersion::V1 {
+    return None;
+  }
+
+  // Check marker.
+  let marker: &[u8] = data.get(1..4)?;
+  if marker != DID_MESSAGE_MARKER {
+    return None;
+  }
+
+  // Decode data.
+  let encoding: DIDMessageEncoding = DIDMessageEncoding::try_from(*data.get(4)?).ok()?;
+  let inner: &[u8] = data.get(5..)?;
+  let mut resource: T = match encoding {
+    DIDMessageEncoding::Json => T::from_json_slice(inner),
+    DIDMessageEncoding::JsonBrotli => T::from_json_slice(&compression_brotli::decompress_brotli(inner).ok()?),
+  }
+  .ok()?;
+  resource.set_message_id(message_id);
+  Some(resource)
+}
+
+/// Encodes the data and prepends the current message version and encoding flags to it.
+pub(crate) fn pack_did_message<T: ToJson>(data: &T, encoding: DIDMessageEncoding) -> Result<Vec<u8>> {
+  // Encode data.
+  let encoded_message_data: Vec<u8> = match encoding {
+    DIDMessageEncoding::Json => data.to_json_vec()?,
+    DIDMessageEncoding::JsonBrotli => compression_brotli::compress_brotli(&data.to_json()?)?,
+  };
+
+  // Prepend flags.
+  let encoded_message_data_with_flags =
+    add_flags_to_message(encoded_message_data, DIDMessageVersion::CURRENT, encoding);
+  Ok(encoded_message_data_with_flags)
+}
+
+/// Prepends the message flags and marker magic bytes to the data in the following order:
+/// `[version, marker, encoding, data]`.
+fn add_flags_to_message(
+  mut data: Vec<u8>,
+  message_version: DIDMessageVersion,
+  encoding: DIDMessageEncoding,
+) -> Vec<u8> {
+  let mut buffer: Vec<u8> = Vec::with_capacity(1 + DID_MESSAGE_MARKER.len() + 1 + data.len());
+  buffer.push(message_version as u8);
+  buffer.extend_from_slice(DID_MESSAGE_MARKER);
+  buffer.push(encoding as u8);
+  buffer.append(&mut data);
+  buffer
+}
```

Comment:
mega nit: One could also write this as:
```suggestion
fn add_flags_to_message(data: Vec<u8>, message_version: DIDMessageVersion, encoding: DIDMessageEncoding) -> Vec<u8> {
  std::iter::once(message_version as u8)
    .chain(DID_MESSAGE_MARKER.into_iter().copied())
    .chain(std::iter::once(encoding as u8))
    .chain(data)
    .collect()
}
```

Reply:
Does `chain` pre-allocate a `Vec` of the exact size needed? The above was to avoid unnecessary memory allocations.

Reply:
Since everyone agrees that the original is good enough, I will merge for now since this PR has been open for a long time.